sdk/core/Azure.Core.Experimental/src/Auth/OAuth2BearerTokenAuthenticationPolicy.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Auth;
+
+/// <summary>
+/// A <see cref="PipelinePolicy"/> that uses an <see cref="ITokenProvider"/>
+/// </summary>
+/// <remarks>
+/// Initializes a new instance of the <see cref="OAuth2BearerTokenAuthenticationPolicy"/> class.
+/// </remarks>
+public class OAuth2BearerTokenAuthenticationPolicy : PipelinePolicy
+{
+    private readonly ITokenProvider _tokenProvider;
+    private readonly IScopedFlowToken _flowContext;
+
+    /// <param name="tokenProvider"></param>
+    /// <param name="contexts"></param>
+    public OAuth2BearerTokenAuthenticationPolicy(ITokenProvider tokenProvider, IEnumerable<IReadOnlyDictionary<string, object>> contexts)
+    {
+        _tokenProvider = tokenProvider;
+        _flowContext = GetContext(contexts, tokenProvider);
+    }
+
+    /// <inheritdoc />
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        Token token;
+        if (message.TryGetProperty(typeof(IScopedFlowToken), out var rawContext) && rawContext is IScopedFlowToken scopesContext)
+        {
+            var context = _flowContext.CloneWithAdditionalScopes(scopesContext.Scopes);
+            token = _tokenProvider.GetAccessToken(context, message.CancellationToken);
+        }
+        else
+        {
+            token = _tokenProvider.GetAccessToken(_flowContext, message.CancellationToken);
+        }
+        message.Request.Headers.Set("Authorization", $"Bearer {token.TokenValue}");
+
+        ProcessNext(message, pipeline, currentIndex);
+        return;
+    }
+
+    /// <inheritdoc />
+    public override ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        return ProcessNextAsync(message, pipeline, currentIndex);
+    }
+
+    internal static IScopedFlowToken GetContext(IEnumerable<IReadOnlyDictionary<string, object>> contexts, ITokenProvider tokenProvider)
```

Comment:
Proposed approach for how the policy interrogates the token provider for the auth flow it supports and then matches that with the context produced by the provider given the supplied contexts from the generated client.

Reply:
nit:  The method name seems a bit odd, in that we're returning a token and not some context type.   Maybe something like `GetTokenForContext`?

Reply:
Renamed the interfaces to replace the `Token` suffix with `Context`.